config.js

```diff
@@ -0,0 +1,79 @@
+import request from 'request';
+import promisify from './promisify';
+import ConsultantError from './consultant-error';
+import {parseKey, keyApplies} from './key-parser';
+import atob from 'atob';
+
+const loadConfig = async (consulHost, prefix, service) => {
+	const prefixUri = prefix ? `${prefix}/` : '';
+	const response = await promisify(cb => request({
+		uri : `${consulHost}/v1/kv/${prefixUri}${service.name}/?recurse=false`,
+		json : true
+	}, cb));
+
+	if (response.statusCode === 404) {
+		const prefixText = prefix ? ` with prefix '${prefix}'` : '';
+		throw new ConsultantError(`'${service.name}'${prefixText} cannot be found in Consul`);
+	}
+	if (response.statusCode !== 200) {
+		throw new ConsultantError(`Error retrieving data from Consul: ${response.statusCode}: ${response.body}`);
+	}
+	return response.body
+		.map(kvPair => {
+			return {key : parseKey(kvPair.Key, prefix), value : kvPair.Value};
+		})
+		.filter(kvPair => kvPair.key.key && keyApplies(kvPair.key, service))
+		.map(kvPair => {
+			return {[kvPair.key.key] : atob(kvPair.value)};
+		})
+		.reduce((properties, kvPair) => Object.assign(properties, kvPair), {});
+};
+
+const updateConfig = (properties, newConfig, callbacks) => {
+	const updates = Object.keys(newConfig).filter(key => properties[key] !== newConfig[key]);
+
+	updates.reduce((props, key) => {
+		props[key] = newConfig[key];
+		return props;
+	}, properties);
+
+	const deletes = Object.keys(properties).filter(key => !newConfig[key]);
+	deletes.forEach(key => delete properties[key]);
+
+	if (updates.length > 0 || deletes.length > 0) {
+		callbacks.forEach(callback => callback(properties))
+	}
+};
+
+export default async function ({consulHost, service, prefix, interval = 500}) {
+	const callbacks = new Set();
+
+	const host = consulHost || process.env.CONSUL_HOST || 'http://localhost:8500';
+
+	service = {
+		name : service.name || process.env.SERVICE_NAME,
+		dataCenter : service.dataCenter || process.env.SERVICE_DC,
+		host : service.host || process.env.SERVICE_HOST,
+		instance : service.instance || process.env.SERVICE_INSTANCE
+	};
+
+	const properties = await loadConfig(host, prefix, service);
+
+	let timerId = 0;
+	if (interval > 0) {
+		timerId = setInterval(async () => await updateConfig(properties, await loadConfig(host, prefix, service), callbacks), interval);
+	}
+
+	return {
+		properties,
```

Comment:
I'm not a fan of a shared properties object. There are no safe guards that other code won't alter this, and it's very easy to do so accidentally in JS. Additionally, it is quite prone to subtle bugs in async code. E.g. consider this seemingly correct snippet:

```js
async function() {
    await fetch(properties.fetchUrl);
    console.log(`Succesfully fetched ${properties.fetchUrl}`);
}
```

If the `fetchUrl` has changed between the call to await and the log then the log statement won't match what actually happened. 

I would strongly suggest to instead expose a getter that creates a new fresh clone on every call. This prevents all these global state issues, and additionally it helps a lot with change detection (it allows users to see *what* has changed, not just *if* something has changed, and it also allows users to do "lazy" change detection). 


Making this a getter function would also eliminate most of the logic in `updateConfig`, since it then only needs to trigger listeners instead of updating global state.